src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBulkRecordReader.java

```diff
@@ -143,50 +170,59 @@ public void close() throws IOException {
+                                                       resultIdIndex, resultIds.length));
+    }
+    InputStream queryResponseStream = bulkConnection.getQueryResultStream(jobId, batchId, resultIds[resultIdIndex]);
+
+    if (resultIdIndex == 0) {
+      CSVFormat csvFormat = CSVFormat.DEFAULT
+        .withHeader()
+        .withQuoteMode(QuoteMode.ALL)
+        .withAllowMissingColumnNames(false);
+      csvParser = CSVParser.parse(queryResponseStream, StandardCharsets.UTF_8, csvFormat);
+      if (csvParser.getHeaderMap().isEmpty()) {
+        throw new IllegalStateException("Empty response was received from Salesforce, but csv header was expected.");
+      }
+      // Read and save the CSV headers from the first query result.
+      csvHeaders = csvParser.getHeaderMap().keySet().toArray(new String[0]);
+    } else {
+      CSVFormat csvFormat = CSVFormat.DEFAULT
+        .withHeader(csvHeaders)
+        .withQuoteMode(QuoteMode.ALL);
+      csvParser = CSVParser.parse(queryResponseStream, StandardCharsets.UTF_8, csvFormat);
     }
-
     parserIterator = csvParser.iterator();
+    resultIdIndex++;
   }
 
   /**
    * Wait until a batch with given batchId succeeds, or throw an exception
    *
    * @param bulkConnection bulk connection instance
-   * @param jobId a job id
-   * @param batchId a batch id
-   * @return an input stream which represents a current batch response, which is a bunch of lines in csv format.
-   *
+   * @return array of batch result ids
    * @throws AsyncApiException  if there is an issue creating the job
    * @throws InterruptedException sleep interrupted
    */
-  public InputStream waitForBatchResults(BulkConnection bulkConnection, String jobId, String batchId)
+  private String[] waitForBatchResults(BulkConnection bulkConnection)
     throws AsyncApiException, InterruptedException {
 
     BatchInfo info = null;
     for (int i = 0; i < SalesforceSourceConstants.GET_BATCH_RESULTS_TRIES; i++) {
-      info = bulkConnection.getBatchInfo(jobId, batchId);
-
-      if (info.getState() == BatchStateEnum.Completed) {
-        QueryResultList list =
-          bulkConnection.getQueryResultList(jobId, batchId);
-        String[] resultIds = list.getResult();
-
-        List<InputStream> streams = new ArrayList<>(resultIds.length);
-        for (String resultId : resultIds) {
-          streams.add(bulkConnection.getQueryResultStream(jobId, batchId, resultId));
+      try {
+        info = bulkConnection.getBatchInfo(jobId, batchId);
+      } catch (AsyncApiException e) {
+        if (i == SalesforceSourceConstants.GET_BATCH_RESULTS_TRIES - 1) {
+          throw e;
         }
+        LOG.warn("Failed to get info for batch {}. Will retry after some time.", batchId, e);
+        continue;
+      }
 
-        return new SequenceInputStream(Collections.enumeration(streams));
+      if (info.getState() == BatchStateEnum.Completed) {
```

Comment:
may be helpful to add a debug message in the else clause which include the batch id and state (and whatever else might be useful), so that the logs have some indication of what the plugin is doing.

Reply:
done